src/libraries/System.Collections.Concurrent/tests/ConcurrentDictionary/ConcurrentDictionaryTests.cs

```diff
@@ -1052,6 +1052,38 @@ public static void TestTryUpdate()
             Assert.True(dict.TryUpdate(1, new Struct16(2, -2), new Struct16(1, -1)), "TestTryUpdate:  FAILED.  TryUpdate failed for non atomic values ( > 8 bytes)");
         }
 
+        [OuterLoop("Runs for several seconds")]
+        [Fact]
+        public void ConcurrentWriteRead_NoTornValues()
+        {
+            var cd = new ConcurrentDictionary<int, KeyValuePair<long, long>>();
+            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(3));
+            Task.WaitAll(
+                Task.Run(() =>
+                {
+                    for (long i = 0; !cts.IsCancellationRequested; i++)
+                    {
+                        cd[0] = new KeyValuePair<long, long>(i, i);
+                    }
+                }),
+                Task.Run(() =>
+                {
+                    while (!cts.IsCancellationRequested)
+                    {
+                        cd.TryGetValue(0, out KeyValuePair<long, long> item);
```

Comment:
Just quickly scanned through the linked comment in #33132, and the behavior that allows this kind of race to happen seems pretty scary.

When an item is set in `cd[0]` by the other thread, a new value is created. `TryGetValue()` should return that value, and `item` should be pointing to the value that was there when `TryGetValue()` looked it up, even if the other thread replaced it. Otherwise, we'll just get silent concurrent bugs that will be painful to find.

Thanks for adding this test.